py-polars/src/lazyframe/mod.rs

```diff
@@ -566,12 +567,42 @@ impl PyLazyFrame {
         Ok((df.into(), time_df.into()))
     }
 
-    fn collect(&self, py: Python) -> PyResult<PyDataFrame> {
+    fn collect(&self, py: Python, lamdba_post_opt: Option<PyObject>) -> PyResult<PyDataFrame> {
         // if we don't allow threads and we have udfs trying to acquire the gil from different
         // threads we deadlock.
         let df = py.allow_threads(|| {
             let ldf = self.ldf.clone();
-            ldf.collect().map_err(PyPolarsErr::from)
+            if let Some(lambda) = lamdba_post_opt {
+                ldf._collect_post_opt(|root, lp_arena, expr_arena| {
+                    Python::with_gil(|py| {
+                        let nt = NodeTraverser::new(
+                            root,
+                            std::mem::take(lp_arena),
+                            std::mem::take(expr_arena),
+                        );
+
+                        // Get a copy of the arena's.
+                        let arenas = nt.get_arenas();
+
+                        // Pass the node visitor which allows the python callback to replace parts of the query plan.
+                        // Remove "cuda" or specify better once we have multiple post-opt callbacks.
+                        lambda.call1(py, (nt,)).map_err(
+                            |e| polars_err!(ComputeError: "'cuda' conversion failed: {}", e),
+                        )?;
```

Comment:
OK so the contract here is that the callback should modify the `NodeTraverser` via `set_udf` to substitute in the subtree it can execute. But that callback isn't allowed to hold on to the `NodeTraverser`, since when the callback actually runs the arenas inside it are garbage.

Reply:
Indeed. Any state that needs to be preserved should be in the callback itself.